app/properties/claim/actions.ts

```diff
@@ -0,0 +1,237 @@
+            },
+            message: 'Invalid Fields Submitted'
+        }
+    }
+    // Need to check more things here before we can claim a property
+
+    const cookieStore = cookies();
+    const supabase = createServerClient(cookieStore);
+
+    // check if the user is logged in
+    const { data: { user }, error: userError } = await supabase.auth.getUser()
+    if (userError)
+        return {
+            message: 'Error Fetching User'
+        };
+
+
+    if (!user)
+        return {
+            message: 'User Not Logged In'
+        };
+
+    //  check a the user is already the landlord of the property for this time period
+    const { data: propertyOwnershipList, error: propertyOwnershipError } = await supabase
+        .from('property_ownership')
+        .select('landlord_id, started_at, ended_at')
+        .eq('property_id', property_id)
+
+    if (propertyOwnershipError)
+        return {
+            message: 'Error Fetching Property Ownership'
+        };
+
+
+    // if empty then good
+    if (propertyOwnershipList.length == 0)
+        return setPropertyOwnership(
+            property_id,
+            landlord_id,
+            started_at,
+            ended_at
+        );
+
+    // If not empty, for each existing property ownership record, there several cases to consider:
+    // 1. The existing claim is closed
+    //    1.1 The existing claim does not overlap with the new claim, continue
+    //    1.2 The existing claim overlaps with the new claim, fail
+    // 2. The existing claim is open 
+    //    2.1 The new claim starts after the existing claim, close the existing claim & continue
+    //    2.2 The new claim starts before the existing claim
+    //        2.2.1 The new claim is closed and ends before the existing claim starts, continue
+    //        2.2.2 The new claim is open or ends after the existing claim starts, fail
+
+    // To handle 2.1, track whether there is an open claim to close as we iterate through the existing claims
+    let openClaimToClose: false | { started_at: Date } = false;
+
+    for (const propertyOwnership of propertyOwnershipList) {
+        const existing_start = new Date(propertyOwnership.started_at)
+        const existing_end = propertyOwnership.ended_at ? new Date(propertyOwnership.ended_at) : new Date()
+
+        // tried to claim the same property again
+        if (propertyOwnership.landlord_id == user.id)
+            return {
+                message: 'User is already the landlord of this property'
+            };
+
+        // new start date is before current end date
+        if (existing_end && ended_at < existing_end)
+            return {
+                message: 'New start date is before the current end date'
+            };
+
+        // new end date is after current start date
+        if (existing_start && started_at > existing_end)
+            return {
+                message: 'New end date is after the current start date'
+            };
+
+    }
```

Comment:
I believe what we are trying to do here is what I outlined in the comment here

I changed the layout of this code block from what you had but I don't believe I changed much if any of the logic outside of moving the adding of ownership to after the loop.

after making it more readable it seems like it does not do what we want (it also doesn't do what you say it does in the comments, your 'new start date is before current end date' check only references end dates).

Can you check whether my test plan in the comments is what we want, and if it is, adjust this section to match it